tcp/src/dispatch.rs

```diff
@@ -84,23 +84,18 @@ impl<S: Stream<Item = String, Error = std::io::Error>> Stream for PeerMessageQue
 
 	fn poll(&mut self) -> Poll<Option<String>, std::io::Error> {
 		// check if we have response pending
-		match self.up.poll() {
-			Ok(Async::Ready(Some(val))) => {
-				return Ok(Async::Ready(Some(val)));
-			}
-			Ok(Async::Ready(None)) => {
-				// this will ensure that this polling will end when incoming i/o stream ends
-				return Ok(Async::Ready(None));
-			}
-			_ => {}
-		}
 
-		match self.receiver.poll() {
-			Ok(result) => Ok(result),
-			Err(send_err) => {
+		match (self.up.poll(), self.receiver.poll()) {
+			(Ok(Async::Ready(Some(v))), _) => Ok(Async::Ready(Some(v))),
```

Comment:
I'm not 100% sure about this, it seems that we might loose some items from `receiver` stream if both are ready.
Perhaps we could introduce simple buffering by introducing an `Option<Poll<...>>` that is checked before polling.

Reply:
Sounds like we want to use this: https://docs.rs/futures/0.1.26/futures/stream/struct.Select.html

Reply:
yeah, I think so the only problem is that the error types, `() vs std::io::Error`

Reply:
one of the errors is ignored anyway (or "should never happen"), so we can just map it to the other.

Reply:
I haven't been able to use `Stream::select` successfully in this case because it requires both streams != NotReady, https://docs.rs/futures/0.1.26/src/futures/stream/select.rs.html#35-61

When I have manually tested it, one channel is `NotReady` and the other one is `active`.
Thus, I'm not sure how to solve it properly. 

A potential `work-around` could be drop it when `ch1 == NotReady && ch2 == Ready(None)` but we then could potentially loose data if ch1 receive data in the future.

Thoughts?

Reply:
I don't understand why both need to be `NotReady`?

Reply:
Sorry for the bad/faulty explanation, I meant that both streams need to return `Ready(None)` for it to be resolved

Reply:
Ahh, I thought the whole purpose of this pr was to fix exactly this :D Then, `Stream::select` is not correct. 
So, we need to look if `receiver` received some messages but don't wait for it to end? Couldn't it be that we get to a point where `receiver` always returns a value, while `up` is already finished? 
That sounds like some weird behavior.

Reply:
true true I agree that we ideally we should use `Stream::select`, guess I got frustrated yesterday let me debug this further